### ISSUE TITLE
[React 18] hydrateRoot(document, <RemixBrowser />) causes app crash with any scripts that modified DOM before hydration

### ISSUE THREAD

**Issue by hrgui** [remix-run/remix#2947]
### What version of Remix are you using?

the main branch of remix (hash a759affafe30c8df27e36f6253a1d9a574ebf18d)

### Steps to Reproduce

Install the Dark Reader Chrome Extension: https://chrome.google.com/webstore/detail/dark-reader/eimadpbcbfnmbkopoojfekhnkhdbieeh?hl=en-US

1. Checkout the remix repo (this repo.)
2. Run `yarn playground:new`
3. After finishing, run `yarn dev` in the newly created playground folder, `yarn watch` in the root of remix 

### Expected Behavior

should see Remix Playground, Sign up / Log In buttons with Dark Reader able to run

### Actual Behavior

The playground loads for a sec, then when React does hydrateRoot, it fails to load if the Dark Reader Chrome Extension is enabled. 

The workaround is to don't use any chrome extensions that modify the DOM in order to run the playground.

```
react-dom.development.js:86 Warning: Expected server HTML to contain a matching <meta> in <head>.
    at meta
    at Meta (http://localhost:3000/build/_shared/chunk-EYMZ6H3Z.js:2816:7)
    at head
    at html
    at App
    at RemixRoute (http://localhost:3000/build/_shared/chunk-EYMZ6H3Z.js:2586:3)
    at Routes2 (http://localhost:3000/build/_shared/chunk-EYMZ6H3Z.js:2569:7)
    at Router (http://localhost:3000/build/_shared/chunk-EYMZ6H3Z.js:679:15)
    at RemixCatchBoundary (http://localhost:3000/build/_shared/chunk-EYMZ6H3Z.js:1079:10)
    at RemixErrorBoundary (http://localhost:3000/build/_shared/chunk-EYMZ6H3Z.js:1004:5)
    at RemixEntry (http://localhost:3000/build/_shared/chunk-EYMZ6H3Z.js:2466:12)
    at RemixBrowser (http://localhost:3000/build/_shared/chunk-EYMZ6H3Z.js:3213:27)
printWarning @ react-dom.development.js:86
5react-dom.development.js:14344 Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.
    at throwOnHydrationMismatch (react-dom.development.js:14344:9)
    at tryToClaimNextHydratableInstance (react-dom.development.js:14372:7)
    at updateHostComponent$1 (react-dom.development.js:20636:5)
    at beginWork (react-dom.development.js:22373:14)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:4157:14)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:4206:16)
    at invokeGuardedCallback (react-dom.development.js:4270:31)
    at beginWork$1 (react-dom.development.js:27243:7)
    at performUnitOfWork (react-dom.development.js:26392:12)
    at workLoopSync (react-dom.development.js:26303:5)

react-dom.development.js:11996 Uncaught DOMException: Failed to execute 'removeChild' on 'Node': The node to be removed is not a child of this node.
    at removeChildFromContainer (http://localhost:3000/build/entry.client-R32SQO3U.js:7970:23)
    at unmountHostComponents (http://localhost:3000/build/entry.client-R32SQO3U.js:16770:17)
    at commitDeletion (http://localhost:3000/build/entry.client-R32SQO3U.js:16814:13)
    at commitMutationEffects_begin (http://localhost:3000/build/entry.client-R32SQO3U.js:16958:19)
    at commitMutationEffects (http://localhost:3000/build/entry.client-R32SQO3U.js:16946:11)
    at commitRootImpl (http://localhost:3000/build/entry.client-R32SQO3U.js:18515:13)
    at commitRoot (http://localhost:3000/build/entry.client-R32SQO3U.js:18446:13)
    at performSyncWorkOnRoot (http://localhost:3000/build/entry.client-R32SQO3U.js:18067:11)
    at flushSyncCallbacks (http://localhost:3000/build/entry.client-R32SQO3U.js:8615:30)
    at commitRootImpl (http://localhost:3000/build/entry.client-R32SQO3U.js:18590:11)
```

**Comment by hrgui** [comment]
Note:

Upgrading react-dom/react to v18 and changing hydrate to hydrateRoot is also affected by this bug. I've provided a test repo to use: https://github.com/hrgui/react-18-remix-app

When doing the following changes:

package.json:
```diff
- "react": "^17.0.2",    
+ "react": "^18.0.0",
- "react-dom": "^17.0.2",    
+ "react-dom": "^18.0.0"
```


entry.client.jsx:
```diff
import { RemixBrowser } from "@remix-run/react";
- import { hydrate } from "react-dom";
+ import { hydrateRoot } from "react-dom/client";

- hydrate(<RemixBrowser />, document);
+ hydrateRoot(document, <RemixBrowser />);
```

**Comment by DAlperin** [comment]
I deployed my first production remix app and have gotten consistent reports from users who have any extension that mutated the DOM (many many of them) I reverted to react 17, but this issue should be considered 'breaking'.

We should probably retitle this issue to reflect that it's not just dark reader, but any extension that injects script tags or mutates the DOM.

**Comment by ksavblik** [comment]
Having the same issue, and agree with @DAlperin for renaming this one
So the real problem is not the mismatch itself, but in the unrecoverable error from new hydrateRoot(in React 17 hydrate just warns about the mismatch)
This is actually relevant for other frameworks as well, not only remix
react issues: [#22833](https://github.com/facebook/react/issues/22833), [#24430](https://github.com/facebook/react/issues/24430)
The question is whether this behavior is expected or not

**Comment by hrgui** [comment]
> We should probably retitle this issue to reflect that it's not just dark reader, but any extension that injects script tags or mutates the DOM.

Changed the title to reflect the true issue.  This also applies for this tweet https://twitter.com/ebey_jacob/status/1509666902477402135

> This is actually relevant for other frameworks as well(that render the whole document), not only remix
react issues: https://github.com/facebook/react/issues/22833, https://github.com/facebook/react/issues/24430

In my non-remix app, I had to do this fix: https://github.com/hrgui/nodejs-react-sc-demo/commit/ac872341d5e2c93d80c562a0f4056e2ae757c7da. 

- The server renders the entire document
- The browser only hydrates the root div element.

---

The question is, how do we make this work with `<RemixBrowser />` ? I tried the following:

```tsx
hydrateRoot(document.body, <RemixBrowser />);
```
^ this doesn't crash, but it causes the app to fall back to client side rendering. This is because there is obviously a hydration mismatch - `<RemixBrowser />` is the entire document.

```tsx
hydrateRoot(document.getElementById('root'), <RemixBrowser />);
```
^ this doesn't crash, but same issue - it causes the app to fall back to client side rendering. This is because there is obviously a hydration mismatch - `<RemixBrowser />` is the entire document.

```tsx
hydrateRoot(document.getElementById('root'), <Outlet />);
```
^ This doesn't present any errors, but client-side hydration doesn't work as a simple counter does not seem to increment.

**Comment by DAlperin** [comment]
> The question is, how do we make this work with <RemixBrowser /> ? I tried the following:

The RemixBrowser API needs a new surface for this use case, I think. I can take a crack at a PR this week if I have time. It would be nice to hear from the core team on how they would like to approach this before I try anything though.

Cc: @mjackson @ryanflorence

**Comment by DAlperin** [comment]
Though, on second thought not being able to hydrate the <head> content would break a lot of things. So I'm not sure that this is the right tradeoff.

**Comment by jacob-ebey** [comment]
Cross linking this: https://github.com/facebook/react/issues/22833

**Comment by Aprillion** [comment]
I got reports that my remix site stopped working after update to React 18 from users with browser plugins which add <meta> or <script> tags, e.g.:

Dark Reader: https://chrome.google.com/webstore/detail/dark-reader/eimadpbcbfnmbkopoojfekhnkhdbieeh
Express VPN: https://chrome.google.com/webstore/detail/expressvpn-vpn-proxy-for/fgddmllnllkalaagkghckoinaemmogpe

Does anyone know of a better workaround than reverting to React 17 until this issue is solved by either React or Remix team?

**Comment by jacob-ebey** [comment]
I believe the React team will have a fix for this in the next release. The [18.2.0-next-a412d787e-20220518](https://www.npmjs.com/package/react/v/18.2.0-next-a412d787e-20220518) release seems to address the underlying issue.

**Comment by hrgui** [comment]
@jacob-ebey I just created a codesandbox with 18.2.0-next-a412d787e-20220518 with the react SSR example.

https://codesandbox.io/s/muddy-worker-t9g9ud
https://t9g9ud.sse.codesandbox.io/

It no longer crashes the application, which is good. However, the entire application still has to fallback to client-side rendering, which I believe will cause double fetching in this case.

**Comment by hrgui** [comment]
From my understanding, 18.2.0-next-a412d787e-20220518 contains the fix  https://github.com/facebook/react/pull/24523. The expectations set there is:  yes, there will be hydration mismatches, but the client will fallback.

In that case, the example in https://t9g9ud.sse.codesandbox.io/ is working as per the react team intended for now.

As per https://github.com/reactjs/reactjs.org/issues/4656#issuecomment-1126180807, I believe the react team is working on a bigger plan to make this work on a more resilient way.

**Comment by ryanyogan** [comment]
If you upgrade to https://www.npmjs.com/package/react/v/18.2.0-next-a412d787e-20220518 you can do the following:

In your `entry.client.tsx` file

```
import { RemixBrowser } from "@remix-run/react";
import { hydrateRoot } from "react-dom/client";

hydrateRoot(document, <RemixBrowser />);
```

In your `entry.server.tsx` file

```
import { PassThrough } from "stream";
import { renderToPipeableStream } from "react-dom/server";
import { RemixServer } from "@remix-run/react";
import { Response } from "@remix-run/node";
import type { EntryContext, Headers } from "@remix-run/node";
import isbot from "isbot";

const ABORT_DELAY = 5000;

export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {
  const callbackName = isbot(request.headers.get("user-agent"))
    ? "onAllReady"
    : "onShellReady";

  return new Promise((resolve, reject) => {
    let didError = false;

    const { pipe, abort } = renderToPipeableStream(
      <RemixServer context={remixContext} url={request.url} />,
      {
        [callbackName]() {
          let body = new PassThrough();

          responseHeaders.set("Content-Type", "text/html");

          resolve(
            new Response(body, {
              status: didError ? 500 : responseStatusCode,
              headers: responseHeaders,
            })
          );
          pipe(body);
        },
        onShellError(err: unknown) {
          reject(err);
        },
        onError(error: unknown) {
          didError = true;
          console.error(error);
        },
      }
    );
    setTimeout(abort, ABORT_DELAY);
  });
}
```

And you should be streaming from the server again on the initial render.  

This is just a short-term hack; I pulled this from Kent Dodds's upcoming Front End Masters course, highly recommend everyone attend those!

**Comment by knpwrs** [comment]
React 18.2.0 is `latest` as of June 14 which includes https://github.com/facebook/react/pull/24523

**Comment by 0Lucifer0** [comment]
it's also preventing loom (extension) to work as the extension add a <loom-container> to the dom. Seems like it just break any extension that update the dom and having a workaround every potential extension is going to be a nightmare

fix for loom
```
document.querySelectorAll('loom-container').forEach((s) => {
    s.parentNode.removeChild(s);
});
```

**Comment by aaacafe786** [comment]
I would like to bump this issue, we have upgraded our stack to the latest version of react (18.2.0), we modified entry.server (https://github.com/remix-run/indie-stack/blob/main/app/entry.server.tsx) and entry.client (https://github.com/remix-run/indie-stack/blob/main/app/entry.client.tsx) inline with the latest version of the indie stack.

After deploying to production, we are finding errors:
![image](https://user-images.githubusercontent.com/102602405/181050819-aebf8550-a129-4b5f-bd14-e6c4a89e1705.png)

If you would like to see a reproduction of the issue, please see here: https://pr-29-osc-academic-hub.fly.dev/

Obviously this diverges slightly from the original issue, as this screenshot demonstrates the error in incognito mode with no extensions installed.

**Comment by cjoecker** [comment]
I realice that the issue comes from the `<Scripts />` component. But still I'm not sure why.

**Comment by bangngo52** [comment]
I have the same experience with @cjoecker when removing the <Script /> the message is disappeared

**Comment by CanRau** [comment]
That's because without `<Scripts/>` you're site is completely static, no JavaScript, no hydration and therefore no client hydration, so no hydration errors

**Comment by dr3** [comment]
cc @aaacafe786 

It appears this issue is a documented "gotcha" https://remix.run/docs/en/v1/pages/gotchas#browser-extensions-injecting-code

With the solution being "Check out the page in incognito mode, the warning should disappear."
Granted this feels like a very much a non-solution, as the error happens in production, and affects any user who has certain (popular) chrome extensions installed

@hrgui Should this issue be closed now (primary issue was fixed in react 18.2.0, and we can open a new issue to track the "gotcha"?

**Comment by hrgui** [comment]
> Check out the page in incognito mode, the warning should disappear.

Note that users can still use extensions in incognito if they desire (just need to check the checkbox)

To be fair, it looks like the original problem has been solved as the application no longer crashes. However, the application resorts to client-side rendering as of React 18.2.0 in the case when an injection happens from an extension.

@dr3 Is there a new issue already opened for the gotcha? Once that new issue is opened, then I don't mind closing this issue as solved and encouraging folks to move on to the gotcha issue.

**Comment by dr3** [comment]
I have since found [this comment](https://github.com/remix-run/remix/issues/1077#issuecomment-1117457264) from @kentcdodds where he says the issue isn't something that remix can fix. 

I have found [this issue in the react repo](https://github.com/facebook/react/issues/24430) which seems to be the same issue

**Comment by hrgui** [comment]
Sounds good @dr3, I have closed this issue as completed.

The original problem of an application crash has been fixed in React 18.2.0. 

Users that have chrome extensions that do inject into the DOM will encounter a client-side re-render warning. Follow https://github.com/facebook/react/issues/24430. 

I believe the only remaining action the Remix team could possibly do is not hydrate the entire document in order to fix the problem. Otherwise, we are waiting on https://github.com/facebook/react/issues/24430.

**Comment by iamvijaydev** [comment]
Hi, a dev from 2024, who also faced the same issue with Cloudflare Pages template. I tested it in a variety of configurations. Even with all the extensions disabled. As someone pointed out above (/previously closed thread) the core problem is with `react@18.2`.

After trying a few React builds the following canary build solved the issue. This also worked with plugins like Grammarly, which we know changes the DOM before hydration.

```diff
- "react": "^18.2.0",
- "react-dom": "^18.2.0"
+ "react": "19.0.0-canary-33a32441e9-20240418",
+ "react-dom": "19.0.0-canary-33a32441e9-20240418"
```

I hope this will help out there anyone unstuck and continue to use the awsome features of Remix.

**Comment by theghostyced** [comment]
> out there anyone unstuck and continue to use the awsome features of Remix.

Thanks a lot @iamvijaydev
This solved the issue for me man!
